scripts/validate_solver_contract.py

```
@@ -0,0 +1,294 @@
+    testpaths_match = re.search(r"(?ms)^\s*testpaths\s*=\s*(.+?)(?:\n\S|\Z)", text)
+    if testpaths_match and "solvers" in testpaths_match.group(1).split():
+        errors.append("pytest.ini testpaths must not include solvers/")
+    solver_tests = sorted((REPO_ROOT / "tests" / "solvers").glob("test_*.py"))
+    if solver_tests:
+        listed = ", ".join(str(path.relative_to(REPO_ROOT)) for path in solver_tests)
+        errors.append(f"top-level tests/solvers must be empty; use solver-local test.sh: {listed}")
+
+
+def _run_command(command: list[str], *, cwd: Path) -> int:
+    print(f"+ ({cwd.relative_to(REPO_ROOT)}) {' '.join(command)}", flush=True)
+    completed = subprocess.run(command, cwd=cwd, check=False)
```

Ruff (0.15.11)

[error] `subprocess` call: check for execution of untrusted input

(S603)